alacritty_terminal/src/term/mod.rs

```diff
@@ -830,13 +916,138 @@ impl SizeInfo {
+                    };
+
+                    // Get on-screen lines of the selection's locations
+                    let mut start = self.buffer_to_visible(span.start);
+                    let mut end = self.buffer_to_visible(span.end);
+
+                    // Start and end lines are swapped as we switch from buffer to line coordinates
+                    if start > end {
+                        mem::swap(&mut start, &mut end);
+                    }
+
+                    let (start_col, end_col) = if start.line == end.line {
+                        // Trim start/end with partially visible block selection
+                        (max(limit_start, start.col), min(limit_end, end.col))
+                    } else {
+                        // Multiple lines, so do full width damage
+                        (Column(0), cols - 1)
+                    };
+
+                    Some((
+                        Point::new(Line(start.line), start_col),
+                        Point::new(Line(end.line), end_col),
+                    ))
+                },
+                None => None,
+            },
+            None => None,
+        }
+    }
+
+    /// Finalize and retrieve current damage. Note that the user is responsible
+    /// for calling LineDamage::reset on every line returned for efficiency.
+    pub fn get_damage(&mut self) -> &mut Damage {
+        self.damage_cursor();
+
+        self.damage.damage_all |= self.mode.contains(TermMode::INSERT);
+
+        if !self.damage.damage_all {
+            let cols = self.grid.num_cols();
+
+            // Selections
+            //
+            if let Some((upper_left, lower_right)) =
+                self.calculate_selection_damage(cols, &self.grid.selection)
+            {
+                for line in upper_left.line.0..=lower_right.line.0 {
+                    self.damage.expand_line(Line(line), upper_left.col, lower_right.col)
+                }
+            }
+            if let Some((upper_left, lower_right)) =
+                self.calculate_selection_damage(cols, &self.last_selection)
+            {
+                for line in upper_left.line.0..=lower_right.line.0 {
+                    self.damage.expand_line(Line(line), upper_left.col, lower_right.col)
+                }
+            }
+
+            // URL highlights
+            //
+            if let Some((upper_left, lower_right)) =
+                self.calculate_highlight_damage(cols, &self.grid.url_highlight)
+            {
+                for line in upper_left.line.0..=lower_right.line.0 {
+                    self.damage.expand_line(Line(line), upper_left.col, lower_right.col)
+                }
+            }
+            if let Some((upper_left, lower_right)) =
+                self.calculate_highlight_damage(cols, &self.last_highlight)
+            {
+                for line in upper_left.line.0..=lower_right.line.0 {
+                    self.damage.expand_line(Line(line), upper_left.col, lower_right.col)
```

Comment:
There's a bug with this approach. Since you're using `Linear` under the hood, you can write to a non existing line, when your `cols` value is changing (during resize). To repro this issue you should open `tmux`. Type `ls -laF` a bunch of time, type some url, and open it (your launcher should be `firefox --new-window`).

crash log
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:378:21
stack backtrace:
   0:     0x563efd20e714 - backtrace::backtrace::libunwind::trace::h7b0eb74108f5983f
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1:     0x563efd20e714 - backtrace::backtrace::trace_unsynchronized::h1b47ddc501b76c75
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2:     0x563efd20e714 - std::sys_common::backtrace::_print_fmt::hdb3e11118348d382
                               at src/libstd/sys_common/backtrace.rs:77
   3:     0x563efd20e714 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9a91dd8c51296d48
                               at src/libstd/sys_common/backtrace.rs:61
   4:     0x563efd2366cc - core::fmt::write::h0abb1a448222f5be
                               at src/libcore/fmt/mod.rs:1028
   5:     0x563efd20b307 - std::io::Write::write_fmt::heafd3457e1ab0a67
                               at src/libstd/io/mod.rs:1412
   6:     0x563efd210dce - std::sys_common::backtrace::_print::hce8ebf3707f77321
                               at src/libstd/sys_common/backtrace.rs:65
   7:     0x563efd210dce - std::sys_common::backtrace::print::hab52fb573eb13c01
                               at src/libstd/sys_common/backtrace.rs:50
   8:     0x563efd210dce - std::panicking::default_hook::{{closure}}::hb9a3113d679899b0
                               at src/libstd/panicking.rs:188
   9:     0x563efd210ac1 - std::panicking::default_hook::haba400789ba2c5f0
                               at src/libstd/panicking.rs:205
  10:     0x563efd2114cb - std::panicking::rust_panic_with_hook::hf1dd057170b2fcea
                               at src/libstd/panicking.rs:464
  11:     0x563efd21106e - std::panicking::continue_panic_fmt::h05282d9235ed5d3c
                               at src/libstd/panicking.rs:373
  12:     0x563efd210f56 - rust_begin_unwind
                               at src/libstd/panicking.rs:302
  13:     0x563efd23278e - core::panicking::panic_fmt::hf1223493b626f3f5
                               at src/libcore/panicking.rs:139
  14:     0x563efd2326da - core::panicking::panic::h83bebb88d2b6fc94
                               at src/libcore/panicking.rs:70
  15:     0x563efc1ef367 - core::option::Option<T>::unwrap::hf01f3c615c1de07a
                               at /rustc/91fd6283e658e2c7aab2d3f5206fc1891f486af2/src/libcore/macros.rs:41
  16:     0x563efc317117 - alacritty_terminal::term::Damage::expand_line::he683084a0e2ab721
                               at /home/kchibisov/Developer/rust/forks/alacritty-tests/alacritty-damage/alacritty_terminal/src/term/mod.rs:743
  17:     0x563efc317117 - alacritty_terminal::term::Term<T>::get_damage::h132ac6c2f4877493
                               at /home/kchibisov/Developer/rust/forks/alacritty-tests/alacritty-damage/alacritty_terminal/src/term/mod.rs:1022
  18:     0x563efc408308 - alacritty::display::Display::draw::h753271083eb46416
                               at alacritty/src/display.rs:402
  19:     0x563efc23a39e - alacritty::event::Processor<N>::run::{{closure}}::h05839f3ef8d172bb
                               at alacritty/src/event.rs:432
  20:     0x563efc38e7dd - winit::platform_impl::platform::sticky_exit_callback::h107bc9899d46cd05
                               at /home/kchibisov/.cargo/git/checkouts/winit-c2fdb27092aba5a7/6608a02/src/platform_impl/linux/mod.rs:693
  21:     0x563efc4012c5 - winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run_return::h878d158547359256
                               at /home/kchibisov/.cargo/git/checkouts/winit-c2fdb27092aba5a7/6608a02/src/platform_impl/linux/wayland/event_loop.rs:519
  22:     0x563efc38dad6 - winit::platform_impl::platform::EventLoop<T>::run_return::hd2748b69789e6d4f
                               at /home/kchibisov/.cargo/git/checkouts/winit-c2fdb27092aba5a7/6608a02/src/platform_impl/linux/mod.rs:629
  23:     0x563efc46428c - <winit::event_loop::EventLoop<T> as winit::platform::desktop::EventLoopExtDesktop>::run_return::h66d57133fb0bfba2
                               at /home/kchibisov/.cargo/git/checkouts/winit-c2fdb27092aba5a7/6608a02/src/platform/desktop.rs:43
  24:     0x563efc2397dd - alacritty::event::Processor<N>::run::hbd3237a0880a081a
                               at alacritty/src/event.rs:346
  25:     0x563efc24579b - alacritty::run::ha67bce41cc743126
                               at alacritty/src/main.rs:215
  26:     0x563efc2441cb - alacritty::main::h74e8960db0c1205c
                               at alacritty/src/main.rs:106
  27:     0x563efc429210 - std::rt::lang_start::{{closure}}::h82c6786536aa08a5
                               at /rustc/91fd6283e658e2c7aab2d3f5206fc1891f486af2/src/libstd/rt.rs:61
  28:     0x563efd210ef3 - std::rt::lang_start_internal::{{closure}}::hefa3bdc6af77c7a5
                               at src/libstd/rt.rs:48
  29:     0x563efd210ef3 - std::panicking::try::do_call::h1fdd22ace65aadbc
                               at src/libstd/panicking.rs:287
  30:     0x563efd219eba - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:83
  31:     0x563efd211a6d - std::panicking::try::hfb89dbe42a15da49
                               at src/libstd/panicking.rs:265
  32:     0x563efd211a6d - std::panic::catch_unwind::hf2d0ee7f8aef5e8d
                               at src/libstd/panic.rs:396
  33:     0x563efd211a6d - std::rt::lang_start_internal::ha3597cbe02c510ac
                               at src/libstd/rt.rs:47
  34:     0x563efc4291e9 - std::rt::lang_start::hc417976da5e63776
                               at /rustc/91fd6283e658e2c7aab2d3f5206fc1891f486af2/src/libstd/rt.rs:61
  35:     0x563efc245d3a - main
  36:     0x7f62b0a5d153 - __libc_start_main
  37:     0x563efc1d019e - _start
  38:                0x0 - <unknown>

```

However, our selection handling and url handling were changed in the latest master, so keep this crash in mind, when you'll be updating.

Reply:
I suppose this isn't really related to using `Linear` as much as it is related to URL highlights not being constrained to the grid, which was an assumption I had, which based on your trace clearly fail if a resize occurs while a URL is highlighted.

Reply:
I mean, your computing index that is out of bounds, according to my trace.

Reply:
Aha, I missed that the issue was *last* rather than *current* highlight. This makes much more sense. We should simply clear last_* on resize, or possibly ignore them on full damage. The former is probably best to avoid branches.

Reply:
And at this point I recommend you to rebase your PR, since your url highlight thing was changed, and you'll likely will be doing it display.rs now iirc.